### PR TITLE
Add preset save button #118

### PR DESCRIPTION
**Pull request by jandegen**
This PR includes the complete preset feature set.
You can now create and select presets, which are stored inside the backend

image

(The dropdown list is just empty because of not running backend during the screenshot)

### CODE REVIEW

**Review by nicolaskolbenschlag — CHANGES_REQUESTED** [review]
I am unable to run this branch. In `docker-compose up --build` an error occurs. Did not change anything in the configuration. Running the command on other branches works.

**Comment by nicolaskolbenschlag** [comment]
Screenshot from 2022-12-18 15-41-34

**Comment by jandegen** [comment]
> Screenshot from 2022-12-18 15-41-34

This PR is still a draft :) So you probably checked out a work in progress.

It should be fixed now, I only need to fix the test cases

**Comment by nicolaskolbenschlag** [comment]
> 

I see, sorry. I thought you already want it to be reviewd :)

File: Apps/frontend/src/components/DistributeOffsetButton.svelte [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<script>
```
Comment (PhlppKrmr) [inline Apps/frontend/src/components/DistributeOffsetButton.svelte:1]: Why is a previous version of the DistributeButton merged in here? Why not wait for the merge into dev?

File: Apps/frontend/src/components/StartStopButton.svelte [hunk 2]
```diff
@@ -2,25 +2,23 @@
...
-    dispatch("startStop", { buttonValue: hasStarted });
+    dispatch("startStop", { buttonValue: !$channelConfig[channel].enabled });
   };
 </script>
 
-<div class="control-panel--bottom_switch" data-cy="startStopSwitch-{channel}">
+<div class="control-panel--bottom_switch">
   <Switch
     hideLabel
-    small
```
Comment (PhlppKrmr) [inline Apps/frontend/src/components/StartStopButton.svelte:21]: `small` was removed for some reason. Shouldn't impact page all too much, but UI should be a little borken.

File: Apps/frontend/src/components/WaveControls.svelte [hunk 3]
```diff
@@ -0,0 +1,200 @@
+<script>
```
Comment (PhlppKrmr) [inline Apps/frontend/src/components/WaveControls.svelte:1]: Whole file is deprecated in new UI

File: Apps/frontend/src/main.js [hunk 4]
```diff
@@ -1,9 +1,14 @@
 import "./app.scss";
 // @ts-ignore
 import App from "./App.svelte";
+import { channelConfig } from "./stores";
```
Comment (PhlppKrmr) [inline Apps/frontend/src/main.js:4]: Unused import?
Try going into `Apps/frontend` and executing `npm run prettier:fix` and `npm run ling:fix`. The latter should show you unused imports and such.

File: Apps/frontend/src/main.js [hunk 5]
```diff
@@ -1,9 +1,14 @@
...
+
+import { BASE_URL } from "./const";
 
 const app = new App({
   target: document.getElementById("app"),
 });
 
+console.log(`Starting frontend on ${BASE_URL}`);
```
Comment (PhlppKrmr) [inline Apps/frontend/src/main.js:12]: Will logging be available to end user in release?

File: Apps/frontend/src/views/ChannelConfigControlPanel.svelte [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    if (response !== undefined) availableChannelConfigPresets.set(response);
+  }
+
+  function showPopup() {
+    $presetPopupOpen = true;
+    loadAllChannelConfigPresets();
+  }
+</script>
+
+{#if $presetPopupOpen}
```
Comment (PhlppKrmr) [inline Apps/frontend/src/views/ChannelConfigControlPanel.svelte:24]: The popup should probably be the nav, not be inside the nav.

File: Apps/frontend/src/views/ControlPanelSide.svelte [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<script>
```
Comment (PhlppKrmr) [inline Apps/frontend/src/views/ControlPanelSide.svelte:1]: Would prefer `ControlPanelRight`. Am also fine with the current name.

**Review by PhlppKrmr — CHANGES_REQUESTED** [review]
First rough review, might add more later on.
I didn't mention some of the stuff regarding the `DistributeButton` cause it should be fixed when you pull `dev` when the button is merged in there.

File: Apps/frontend/src/components/StartStopButton.svelte [hunk 8]
```diff
@@ -2,25 +2,23 @@
...
 
   const handleStartStop = async () => {
-    hasStarted = !hasStarted;
-    dispatch("startStop", { buttonValue: hasStarted });
+    dispatch("startStop", { buttonValue: !$channelConfig[channel].enabled });
   };
 </script>
 
-<div class="control-panel--bottom_switch" data-cy="startStopSwitch-{channel}">
+<div class="control-panel--bottom_switch">
```
[the remark this answers is not in the document]
Comment (jandegen) [reply Apps/frontend/src/components/StartStopButton.svelte]: Already re-added directly to the switch

File: Apps/frontend/src/components/DistributeOffsetButton.svelte (hunk 1 above)
Comment (jandegen) [reply Apps/frontend/src/components/DistributeOffsetButton.svelte:1]: Old merge from before christmas. Not used

File: Apps/frontend/src/components/StartStopButton.svelte (hunk 2 above)
Comment (jandegen) [reply Apps/frontend/src/components/StartStopButton.svelte:21]: image
 UI seems fine

Comment (PhlppKrmr) [reply Apps/frontend/src/components/StartStopButton.svelte:21]: Yeah, it is. Left is without small, right is with small:
image

File: Apps/frontend/src/main.js (hunk 5 above)
Comment (jandegen) [reply Apps/frontend/src/main.js:12]: Not directly relevant for end users but helps for quick checking

File: Apps/frontend/src/views/ChannelConfigControlPanel.svelte (hunk 6 above)
Comment (jandegen) [reply Apps/frontend/src/views/ChannelConfigControlPanel.svelte:24]: Not sure, what to change, I recreated the logic friom the ControlPanelBottom

File: Apps/frontend/src/views/ControlPanelSide.svelte (hunk 7 above)
Comment (jandegen) [reply Apps/frontend/src/views/ControlPanelSide.svelte:1]: Agree

File: Apps/frontend/src/components/DistributeOffsetButton.svelte (hunk 1 above)
Comment (jandegen) [reply Apps/frontend/src/components/DistributeOffsetButton.svelte]: Deleted

File: Apps/frontend/src/components/StartStopButton.svelte (hunk 2 above)
Comment (jandegen) [reply Apps/frontend/src/components/StartStopButton.svelte:21]: So do we keep it or do we make it small again?

File: Apps/frontend/src/components/WaveControls.svelte (hunk 3 above)
Comment (jandegen) [reply Apps/frontend/src/components/WaveControls.svelte]: Deleted

File: Apps/frontend/src/views/ChannelConfigControlPanel.svelte (hunk 6 above)
Comment (jandegen) [reply Apps/frontend/src/views/ChannelConfigControlPanel.svelte:24]: Let's handle it with refactoring in next sprint

File: Apps/frontend/src/components/StartStopButton.svelte (hunk 2 above)
Comment (jandegen) [reply Apps/frontend/src/components/StartStopButton.svelte:21]: Made it small due to rendering bug

Comment (PhlppKrmr) [reply Apps/frontend/src/components/StartStopButton.svelte:21]: Up to personal preference. If you like it more, keep it.

File: Apps/frontend/src/views/ChannelConfigControlPanel.svelte (hunk 6 above)
Comment (PhlppKrmr) [reply Apps/frontend/src/views/ChannelConfigControlPanel.svelte:24]: Can you ask the POs for a refactoring issue?

**Comment by jandegen** [comment]
> 

I cannot close your conversion.

**Review by motschel123 — APPROVED** [review]
Can't fully test with generator, but the config save&restore works so far without it.
Would be a nice addition to auto hide the view after save/restore

**Review by PhlppKrmr — APPROVED** [review]
Functionality LGTM, some UX not as I'd have expected.

**Pull request merged**